[SPARK-25839] [Core] Implement use of KryoPool in KryoSerializer

patrickbrownsync opened the pull request:
## What changes were proposed in this pull request?

* Implement (optional) use of KryoPool in KryoSerializer, an alternative to the existing implementation of caching a Kryo instance inside KryoSerializerInstance
* Add config key & documentation of spark.kryo.pool in order to turn this on
* Add benchmark KryoSerializerBenchmark to compare new and old implementation
* Add results of benchmark

## How was this patch tested?

Added new tests inside KryoSerializerSuite to test the pool implementation as well as added the pool option to the existing regression testing for SPARK-7766


This is my original work and I license the work to the project under the project’s open source license.

patrickbrownsync committed:
use kryo pool for kryo serializer

patrickbrownsync committed:
fix pool serializable issue

patrickbrownsync committed:
Add option to KryoSerializer to use new KryoPool based implentation
Add tests for new implementation to KryoSerializerSuite.scala
Add benchmark new vs old implemtation in KryoSerializerBenchmark.scala
Add option in Benchmark base for afterAll() shutdown code to facilitate clean benchmark shutdown
Add documentation for spark.kryo.pool configuration option

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

patrickbrownsync committed:
fix typo

xuanyuanking commented on core/src/main/scala/org/apache/spark/serializer/KryoSerializer.scala:
```diff
@@ -92,6 +94,18 @@ class KryoSerializer(conf: SparkConf)
       new KryoOutput(bufferSize, math.max(bufferSize, maxBufferSize))
     }
 
+
```
nit: extra empty line.

xuanyuanking commented on core/src/test/scala/org/apache/spark/serializer/KryoSerializerBenchmark.scala:
```diff
@@ -0,0 +1,90 @@
+    }
+  }
+
+  def createSparkContext(usePool: Boolean): SparkContext = {
```
We add this but not override `getSparkSession` in `SqlBasedBenchmark`, is it because change conf in SparkSession doesn't work for SparkSession?

patrickbrownsync replied:
I'm not sure I understand the question here, this benchmark class doesn't inherit from `SqlBasedBenchmark` it inherits from `BenchmarkBase` which has no `getSparkSession` method.

xuanyuanking commented on core/src/test/scala/org/apache/spark/serializer/KryoSerializerSuite.scala:
```diff
@@ -33,6 +33,7 @@ import org.apache.spark.serializer.KryoTest._
 import org.apache.spark.storage.BlockManagerId
 import org.apache.spark.util.Utils
 
+
```
nit: extra empty line.

xuanyuanking reviewed (commented):
+1 for this and will try in our own folk. Maybe it will have better performance on long running jobs like streaming?

xuanyuanking commented on core/src/main/scala/org/apache/spark/serializer/KryoSerializer.scala:
```diff
@@ -298,30 +312,40 @@ class KryoDeserializationStream(
+      if (cachedKryo != null) {
+        val kryo = cachedKryo
+        /**
+        * As a defensive measure, call reset() to clear any Kryo state that might have
```
nit for the style.

patrickbrownsync replied:
Sorry I'm new to this, what is the specific style issue here?

10110346 replied:
The `*`after the first line must be aligned with the first `*` of the first line.

patrickbrownsync replied:
Thanks fixed

patrickbrownsync committed:
style fixes

patrickbrownsync committed:
fix comment style

srowen commented on core/src/main/scala/org/apache/spark/serializer/KryoSerializer.scala:
```diff
@@ -92,6 +94,16 @@ class KryoSerializer(conf: SparkConf)
+  }
+
+  @transient
+  lazy val pool = new KryoPool.Builder(factory).softReferences.build
```
`private`?

patrickbrownsync replied:
This is used directly by KryoSerializerInstance

srowen commented on core/src/main/scala/org/apache/spark/serializer/KryoSerializer.scala:
```diff
@@ -92,6 +94,16 @@ class KryoSerializer(conf: SparkConf)
     }
 
+  @transient
+  private lazy val factory: KryoFactory = new KryoFactory() {
```
I think you're welcome to just write ...
```
private lazy val factory = new KryoFactory() {
  override def create: Kryo = newKryo()
}
```
but it doesn't matter.

patrickbrownsync replied:
Thanks, I'll remove, I had a problem previously but it looks like the @transient on the pool is enough itself.

srowen commented on core/src/test/scala/org/apache/spark/serializer/KryoSerializerSuite.scala:
```diff
@@ -456,9 +458,63 @@ class KryoSerializerSuite extends SparkFunSuite with SharedSparkContext {
-    test(s"instance reuse with autoReset = $autoReset, referenceTracking = $referenceTracking") {
-      testSerializerInstanceReuse(autoReset = autoReset, referenceTracking = referenceTracking)
+  for {
+    referenceTracking <- Set(true, false)
```
Not that it matters, but I think this should have originally been Seq not Set

patrickbrownsync replied:
Might as well fix while I'm here

srowen commented on core/src/test/scala/org/apache/spark/serializer/KryoSerializerSuite.scala:
```diff
@@ -456,9 +458,63 @@ class KryoSerializerSuite extends SparkFunSuite with SharedSparkContext {
+  }
+
+  test("SPARK-25839 KryoPool implementation works correctly in multi-threaded environment") {
+    import java.util.concurrent.Executors
```
I'd import at the top of the file.

srowen commented on core/src/main/scala/org/apache/spark/serializer/KryoSerializer.scala:
```diff
@@ -30,6 +30,7 @@ import scala.util.control.NonFatal
 import com.esotericsoftware.kryo.{Kryo, KryoException, Serializer => KryoClassSerializer}
 import com.esotericsoftware.kryo.io.{Input => KryoInput, Output => KryoOutput}
 import com.esotericsoftware.kryo.io.{UnsafeInput => KryoUnsafeInput, UnsafeOutput => KryoUnsafeOutput}
+import com.esotericsoftware.kryo.pool._
```
I'd spell out the imports for clarity unless it's going to run more than 2 lines or something

srowen commented on core/src/test/scala/org/apache/spark/serializer/KryoSerializerSuite.scala:
```diff
@@ -456,9 +458,63 @@ class KryoSerializerSuite extends SparkFunSuite with SharedSparkContext {
+    test(s"instance reuse with autoReset = $autoReset, referenceTracking = $referenceTracking" +
+      s", usePool = $usePool") {
+      testSerializerInstanceReuse(
+        autoReset = autoReset, referenceTracking = referenceTracking, usePool = usePool)
```
Likewise i see this is just how the code was written before but the `foo = foo` style isn't adding anything IMHO. Feel free to not name args

srowen commented on core/src/main/scala/org/apache/spark/serializer/KryoSerializer.scala:
```diff
@@ -84,6 +85,7 @@ class KryoSerializer(conf: SparkConf)
   private val avroSchemas = conf.getAvroSchema
   // whether to use unsafe based IO for serialization
   private val useUnsafe = conf.getBoolean("spark.kryo.unsafe", false)
+  private val usePool = conf.getBoolean("spark.kryo.pool", false)
```
Yep, I'd leave this undocumented

patrickbrownsync replied:
I  had already added this to docs/configuration.md should I remove it?

srowen replied:
I would not document it. This is just a safety valve. In theory, there's no reason to disable this nor would a caller know why to disable it.

patrickbrownsync replied:
Okay, if I'm understanding you correctly I should set the default to true, and remove the documentation. I will go ahead and do that, if I misunderstood let me know. Thanks

srowen replied:
Yeah I think that's a fine position to take, if we can't think of a reason to disable it other than the theoretical unknown unknown bug out there.

patrickbrownsync replied:
Sounds good and done

srowen reviewed (commented):
Looks good to me, I just have some minor questions and suggestions about code style

patrickbrownsync commented:
@srowen Thanks for taking the time to give me feedback! I'll get this stuff fixed

patrickbrownsync committed:
implement import order fixes, style fixes, remove uneeded @transient

patrickbrownsync committed:
default using KryoPool to true, and remove from documentation

SparkQA commented:
**[Test build #4410 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/4410/testReport)** for PR 22855 at commit [`3433622`](https://github.com/apache/spark/commit/3433622d92957607ca3a4f83f199f86f52077627).

SparkQA commented:
**[Test build #4410 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/4410/testReport)** for PR 22855 at commit [`3433622`](https://github.com/apache/spark/commit/3433622d92957607ca3a4f83f199f86f52077627).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

srowen commented:
Oops, something to do with the benchmark class:

```
[error] /home/jenkins/workspace/NewSparkPullRequestBuilder/core/src/test/scala/org/apache/spark/serializer/KryoSerializerBenchmark.scala:41: object creation impossible, since method runBenchmarkSuite in class BenchmarkBase of type (mainArgs: Array[String])Unit is not defined
[error] object KryoSerializerBenchmark extends BenchmarkBase {
[error]        ^
[error] /home/jenkins/workspace/NewSparkPullRequestBuilder/core/src/test/scala/org/apache/spark/serializer/KryoSerializerBenchmark.scala:45: method runBenchmarkSuite overrides nothing.
[error] Note: the super classes of object KryoSerializerBenchmark contain the following, non final members named runBenchmarkSuite:
[error] def runBenchmarkSuite(mainArgs: Array[String]): Unit
[error]   override def runBenchmarkSuite(): Unit = {
[error]                ^
```

srowen commented:
Ping @patrickbrownsync -- with this fixed I think we can merge it

patrickbrownsync committed:
update KryoSerializerBenchmark to conform to new BenchmarkBase

patrickbrownsync commented:
@srowen should be fixed now, I rebased off master and changed KryoSerializerBenchmark to conform to how BenchmarkBase had changed on master

srowen commented on core/src/test/scala/org/apache/spark/serializer/KryoSerializerSuite.scala:
```diff
@@ -431,9 +434,11 @@ class KryoSerializerSuite extends SparkFunSuite with SharedSparkContext {
 
-  private def testSerializerInstanceReuse(autoReset: Boolean, referenceTracking: Boolean): Unit = {
+  private def testSerializerInstanceReuse(
+     autoReset: Boolean, referenceTracking: Boolean, usePool: Boolean): Unit = {
```
Total nit: there are a couple continuation indents that are 3 rather than 4 spaces. Don't bother, unless you need to make other updates.

srowen commented on core/src/main/scala/org/apache/spark/serializer/KryoSerializer.scala:
```diff
@@ -298,30 +309,40 @@ class KryoDeserializationStream(
+      if (cachedKryo != null) {
+        val kryo = cachedKryo
+        /**
+         * As a defensive measure, call reset() to clear any Kryo state that might have
```
Another total nit, not worth touching unless you make other changes, but this is scaladoc-style. Multi-line comments are often just commented with `//` as before.

srowen reviewed (approved):
OK to me. Let me leave it open a day

patrickbrownsync committed:
stlying fixes

patrickbrownsync commented:
@srowen I went ahead and fixed those issues where I could find them, mostly for my own learning experience. Thanks for taking the time to review this many times, really appreciate it

SparkQA commented:
**[Test build #4415 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/4415/testReport)** for PR 22855 at commit [`9069768`](https://github.com/apache/spark/commit/90697681ea39943b130977d7687757ce6a4db722).

SparkQA commented:
**[Test build #4415 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/4415/testReport)** for PR 22855 at commit [`9069768`](https://github.com/apache/spark/commit/90697681ea39943b130977d7687757ce6a4db722).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

patrickbrownsync committed:
re-add needed @transient to the KryoFactory

SparkQA commented:
**[Test build #4417 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/4417/testReport)** for PR 22855 at commit [`60310c0`](https://github.com/apache/spark/commit/60310c0e18613f0c32f19b73e6ac25a49ba25e86).

SparkQA commented:
**[Test build #4417 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/4417/testReport)** for PR 22855 at commit [`60310c0`](https://github.com/apache/spark/commit/60310c0e18613f0c32f19b73e6ac25a49ba25e86).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

patrickbrownsync committed:
Fix failing tests:
KryoSerializerInstance is now lazy and will not allocate a Kryo instance until serializer is called, so update test in KryoSerializerSuite to reflect that
The Serializer method setDefaultClassLoader updates the class loader, in this case we need to reset the KryoPool to reflect the new class loader, unfortunately Kryo does not expose any method of doing this on the pool, so we need to re-create the KryoPool

srowen commented on core/src/main/scala/org/apache/spark/serializer/KryoSerializer.scala:
```diff
@@ -92,6 +94,20 @@ class KryoSerializer(conf: SparkConf)
+    }
+  }
+
+  @transient
```
Back on this item now, if this is no longer a lazy val, and it's transient, how does it get set again after this object itself is serialized? I'm mostly wondering if this is required now.

patrickbrownsync replied:
Yes, you are correct, I don't think this will work

srowen commented on core/src/main/scala/org/apache/spark/serializer/KryoSerializer.scala:
```diff
@@ -214,8 +230,14 @@ class KryoSerializer(conf: SparkConf)
   }
 
+  override def setDefaultClassLoader(classLoader: ClassLoader): Serializer = {
+    defaultClassLoader = Some(classLoader)
```
You can write `super.setDefaultClassLoader(classLoader)` here to inherit the behavior rather than duplicate it. It is just one line now, yes.

Where is defaultClassLoader used in this implementation though? I wonder why it matters that you call `getPool` after this field is set. And if it does, isn't setting it in line 105 not actually helping?

patrickbrownsync replied:
defaultClassLoader is used in newKryo.

I called `getPool` after setting the defaultClassLoader to make sure we don't accidently create a newKryo before the defaultClassLoader is updated. Setting it on 105 was because I don't believe `setDefaultClassLoader` is required to be called.

The issue that I unfortunately didn't notice until these tests failed is that the tests specify that you can `setDefaultClassLoader` after serializing an object (I'm not sure this functionality is actually used), causing an "incorrect" Kryo instance to be in the pool. Unfortunately the pool doesn't expose a way to clear itself out, thus the var, which clearly doesn't work.

I will work on a solution

patrickbrownsync committed:
create internal wrapper around KryoPool to implement reset functionality

wangyum commented on core/src/test/scala/org/apache/spark/serializer/KryoSerializerBenchmark.scala:
```diff
@@ -0,0 +1,90 @@
+ *      Results will be written to "benchmarks/KryoSerializerBenchmark-results.txt".
+ * }}}
+ */
+object KryoSerializerBenchmark extends BenchmarkBase {
```
cc @dongjoon-hyun for Benchmark change.

patrickbrownsync commented on core/src/main/scala/org/apache/spark/serializer/KryoSerializer.scala:
```diff
@@ -92,6 +94,36 @@ class KryoSerializer(conf: SparkConf)
+    }
+  }
+
+  private class PoolWrapper extends KryoPool {
```
This wrapper can be removed when Kryo is updated to 5.0 as the new pool implementation exposes the needed method to clear the pool

patrickbrownsync commented on core/src/main/scala/org/apache/spark/serializer/KryoSerializer.scala:
```diff
@@ -92,6 +94,16 @@ class KryoSerializer(conf: SparkConf)
     }
 
+  @transient
+  private lazy val factory: KryoFactory = new KryoFactory() {
```
This was needed

srowen commented on core/src/main/scala/org/apache/spark/serializer/KryoSerializer.scala:
```diff
@@ -101,13 +101,29 @@ class KryoSerializer(conf: SparkConf)
+    override def run[T](kryoCallback: KryoCallback[T]): T = pool.run(kryoCallback)
+
+    def reset(): Unit = {
+      pool = getPool
```
OK. We could look at updating to Kryo 5 in Spark 3, too, if there were other benefits.
Does the old pool need to get freed or anything here?

patrickbrownsync replied:
The KryoPool interface exposes no way to free it, so I believe that is not necessary

srowen replied:
OK, and it only gets recreated when a new classloader is set, it seems, so it is rare.

SparkQA commented:
**[Test build #4421 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/4421/testReport)** for PR 22855 at commit [`3bfc4eb`](https://github.com/apache/spark/commit/3bfc4ebbf214b6b0fadbaa10aa832303a59de97d).

SparkQA commented:
**[Test build #4421 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/4421/testReport)** for PR 22855 at commit [`3bfc4eb`](https://github.com/apache/spark/commit/3bfc4ebbf214b6b0fadbaa10aa832303a59de97d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

srowen commented:
Merged to master
